### PR TITLE
Docker Swarm SD: Support tasks and service without published ports

### LINKED ISSUE #7684: Docker Swarm: Discover tasks and services exposed in overlay networks
Hello there,

**What did you do?**
I deployed a Prometheus instance on a Docker Swarm to try out the new Docker Swarm Service Discovery.

Here is the docker-compose.yml :
```
version: '3.7'

services:
  prometheus:
    image: prom/prometheus:master # Will change to v2.20.0
    deploy:
      resources:
        limits:
          memory: 1G
        reservations:
          memory: 128M
      labels:
        - "prometheus.enable.service=true"
    stop_grace_period: 1m30s
    ports:
      - published: 9090
        target: 9090
        mode: host
    command:
      - '--log.level=debug'
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--storage.tsdb.retention.time=${PROMETHEUS_RETENTION:-24h}'
    volumes:
      - /data/prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
      - /var/run/docker.sock:/var/run/docker.sock
    user: root
    networks:
      - default
```

I get services and tasks listed in the Service Discovery on http://localhost:9090/service-discovery, including Prometheus, which is kept as a target, while everything else is dropped (as expected).

**What did you expect to see?**
If I deploy the following stack, it does not appear in the new discovered tasks or services.
```
version: '3.7'

services:
  helloworld:
    image: alpine:3.12
    tty: true
    networks:
      - default
```
Same goes for :
```
version: '3.7'

services:
  mysql1:
    image: mysql/mysql-server:8.0.19
    stop_grace_period: 60s
    environment:
      - MYSQL_ROOT_PASSWORD=<password>
      - MYSQL_ROOT_HOST=%
    hostname: server-1
    networks:
      - default
```
However, some more complex stacks are discovered as expected.

Starting Prometheus before or after deploying the stack provides the same result.

**What did you see instead? Under which circumstances?**

**Environment**
Tested on a Swarm with a single node, and a Swarm with 3 master nodes and multiple workers.
Docker version on single master : Docker version 19.03.12, build 48a66213fe
Docker version on multi master : Docker version 19.03.2, build 6a30dfc

* System information:
Single master : Linux 5.3.0-62-generic x86_64
Multi master : Linux 3.10.0-1062.1.1.el7.x86_64 x86_64

* Prometheus version:
```
prometheus, version 2.20.0 (branch: master, revision: 348ff4285ffa59907c9d7fd7eb3cb7d748f42758)
  build user:       root@f7960ff9f1b5
  build date:       20200728-04:45:57
  go version:       go1.14.6
```

* Prometheus configuration file:
```
global:
  scrape_interval:     15s # By default, scrape targets every 15 seconds.
  evaluation_interval: 15s # By default, scrape targets every 15 seconds.
  scrape_timeout: 15s

  external_labels:
    monitor: 'prometheus-monitor-1.prod'

rule_files:
  - "rules/*.rules"

scrape_configs:
  - job_name: 'swarm-services'
    dockerswarm_sd_configs:
      - host: unix:///var/run/docker.sock
      # - host: http://dockerd:2375
        role: services
    relabel_configs:
      # Only keep containers that have a `prometheus-enable-service` label.
      - source_labels: [__meta_dockerswarm_service_label_prometheus_enable_service]
        regex: .+
        action: keep
      # Use the task labels that are prefixed by `prometheus-`.
      - regex: __meta_dockerswarm_service_label_prometheus_(.+)
        action: labelmap
        replacement: $1

  - job_name: 'swarm-tasks'
    dockerswarm_sd_configs:
      - host: unix:///var/run/docker.sock
      # - host: http://dockerd:2375
        role: tasks  
    relabel_configs:                                
      # Only keep containers that should be running.
      - source_labels: [__meta_dockerswarm_task_desired_state]
        regex: running
        action: keep
      # Only keep containers that have a `prometheus-enable-task` label.
      - source_labels: [__meta_dockerswarm_service_label_prometheus_enable_task]
        regex: .+
        action: keep
      # Use the task labels that are prefixed by `prometheus-`.
      - regex: __meta_dockerswarm_service_label_prometheus_(.+)
        action: labelmap
        replacement: $1
```

* Logs:
For reference, even though there's nothing abnormal here.
```
level=info ts=2020-07-28T15:11:48.626Z caller=main.go:345 msg="Starting Prometheus" version="(version=2.20.0, branch=master, revision=348ff4285ffa59907c9d7fd7eb3cb7d748f42758)"
level=info ts=2020-07-28T15:11:48.626Z caller=main.go:346 build_context="(go=go1.14.6, user=root@f7960ff9f1b5, date=20200728-04:45:57)"
level=info ts=2020-07-28T15:11:48.626Z caller=main.go:347 host_details="(Linux 5.3.0-62-generic #56-Ubuntu SMP Tue Jun 23 11:20:52 UTC 2020 x86_64 5efe78907e8f (none))"
level=info ts=2020-07-28T15:11:48.626Z caller=main.go:348 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2020-07-28T15:11:48.626Z caller=main.go:349 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-07-28T15:11:48.627Z caller=main.go:683 msg="Starting TSDB ..."
level=info ts=2020-07-28T15:11:48.627Z caller=web.go:525 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-07-28T15:11:48.631Z caller=head.go:642 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2020-07-28T15:11:48.631Z caller=head.go:656 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=1.288µs
level=info ts=2020-07-28T15:11:48.631Z caller=head.go:662 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2020-07-28T15:11:48.631Z caller=head.go:714 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
level=info ts=2020-07-28T15:11:48.631Z caller=head.go:717 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=13.106µs wal_replay_duration=140.657µs total_replay_duration=166.48µs
level=info ts=2020-07-28T15:11:48.632Z caller=main.go:699 fs_type=EXT4_SUPER_MAGIC
level=info ts=2020-07-28T15:11:48.632Z caller=main.go:700 msg="TSDB started"
level=debug ts=2020-07-28T15:11:48.632Z caller=main.go:701 msg="TSDB options" MinBlockDuration=2h MaxBlockDuration=144m MaxBytes=0B NoLockfile=false RetentionDuration=1d WALSegmentSize=0B AllowOverlappingBlocks=false WALCompression=true
level=info ts=2020-07-28T15:11:48.632Z caller=main.go:819 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
level=debug ts=2020-07-28T15:11:48.632Z caller=manager.go:226 component="discovery manager scrape" msg="Starting provider" provider=*dockerswarm.SDConfig/0 subs=[swarm-services]
level=debug ts=2020-07-28T15:11:48.632Z caller=manager.go:226 component="discovery manager scrape" msg="Starting provider" provider=*dockerswarm.SDConfig/1 subs=[swarm-tasks]
level=info ts=2020-07-28T15:11:48.632Z caller=main.go:847 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml
level=info ts=2020-07-28T15:11:48.632Z caller=main.go:655 msg="Server is ready to receive web requests."
```

Thank you !

### PR DESCRIPTION
**Pull request by roidelapluie**
Mimics k8s discovery behaviour.

Closes #7684 

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -459,7 +459,10 @@ One of the following roles can be configured to discover targets:
 
 #### `services`
 
-The `services` role is used to discover [Swarm services](https://docs.docker.com/engine/swarm/key-concepts/#services-and-tasks).
+The `services` role discovers all [Swarm services](https://docs.docker.com/engine/swarm/key-concepts/#services-and-tasks)
+and exposes their ports as targets. For each published port of a service, a
+single target is generated. If a service has no published ports, a target per
+service is created with the `port` parameter defined in the SD configuration.
```
Comment (brian-brazil) [inline docs/configuration/configuration.md:465]: s/with the/using the/

**Comment by darkl0rd** [comment]
Can confirm this works - however, I did notice that under status => service discovery it shows xx/total count of containers.

For example, 8/1742. Whereas under targets it (correctly) shows 8/8.

**Comment by roidelapluie** [comment]
On 28 Aug 02:23, darkl0rd wrote:
> Can confirm this works - however, I did notice that under status => service discovery it shows xx/total count of containers.
> 
> For example, 8/1742. Whereas under targets it (correctly) shows 8/8.


This is correct, it means that we discovered 1742 targets but only 8
were kept in the final config (1736 were dropped by relabel configs).


-- 
 (o-    Julien Pivotto
 //\    Open-Source Consultant
 V_/_   Inuits - https://www.inuits.eu

**Comment by darkl0rd** [comment]
@roidelapluie If a target is discoverable through multiple paths (overlay networks) it is added for each. What is the preferred way of limiting this? Is there a way to only perform discovery on a specific network?

**Comment by roidelapluie** [comment]
https://prometheus.io/docs/guides/dockerswarm/#scraping-metrics-via-a-certain-network-only

It makes more sense to ask questions like this on the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users) rather than in a GitHub issue. On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.

**Comment by darkl0rd** [comment]
My apologies, I completely missed that piece of documentation and reported it here as I assumed it was something that had not been considered yet.

**Pull request merged**